metabase/metabase
# feat(sdk): use rtk query for loading static questions

Closes https://github.com/metabase/metabase/issues/49620

### Description

Uses rtk-query hooks (e.g. `useGetCardQueryQuery`) to load static questions instead of calling methods in the `CardApi` object manually. There are two primary objectives for this:

1. Simplifies the logic of fetching static questions, and improve consistency with the rest of the codebase which uses rtk-query.
2. Handles request cancellation automatically, without needing to muck around with `AbortController`.

### How to verify

To verify that rtk-query correctly handles request cancellation using Shoppy:

- In your Metabase directory, build the SDK locally with `yarn build-embedding-sdk`
- In your Shoppy directory, link the locally built package and run the app with `rm -rf node_modules/.vite && rm -rf node_modules/@metabase/ && yarn add file:../metabase/resources/embedding-sdk && env NODE_OPTIONS=--max-old-space-size=7000 yarn dev`
  - Update the `file:` path in the above command to match your workspace's directory structure.
- Go to the Product List page (e.g. `http://localhost:3004/admin/products`) on the ProficiencyLabs theme, as that theme has a lot of products to test request cancellation
- On page load, you should now see 6 card requests ("/card/161") and 6 query requests because we uses an IntersectionObserver to only load the component that is rendered. Refer to the below screenshot.

![CleanShot 2568-01-16 at 18 38 01@2x](https://github.com/user-attachments/assets/c5a5af8a-249e-491a-80a3-269ab32067e3)

- Now scroll really fast to the end of the page, and immediately click on the other theme. You should now see a bunch of ongoing requests in the network tab with "X" on it, and a lot more new requests has been made. That means the requests from the previous theme has been cancelled, and more requests from the current theme has been issued.

![CleanShot 2568-01-16 at 18 39 33@2x](https://github.com/user-attachments/assets/ee248e22-78c3-4df0-b0cc-cfd4ea7bfabb)

To verify that StaticQuestion continues to work:

- Questions should continue to load
- SQL question parameters should continue to work
- Updating static visualization types (with `withChartTypeSelector` prop on `StaticQuestion`) should continue to work

### Checklist
